pkg/operator/resource/resourceapply/admissionregistration.go

```diff
@@ -0,0 +1,185 @@
+//
+// To ensure an update in response to state external to the resource spec, the caller should
+// set an annotation representing that external state e.g.
+//
+//   `myoperator.openshift.io/config-resource-version: <resourceVersion>`
+//
+// An update will be performed if:
+//
+// - The required resource metadata differs from that of the existing resource.
+//   - The difference will be detected by comparing the name, namespace, labels and
+//   annotations of the 2 resources.
+//
+// - The generation expected by the operator differs from generation of the existing
+// resource.
+//   - This is the likely result of an actor other than the operator updating a resource
+//   managed by the operator.
+//
+// - The spec of the required resource differs from the spec of the existing resource.
+//   - The difference will be detected via metadata comparison since the hash of the
+//   resource's spec will be set as an annotation prior to comparison.
+
+const genericCABundleInjectorAnnotation = "service.beta.openshift.io/inject-cabundle"
+
+// ApplyMutatingWebhookConfiguration ensures the form of the specified
+// mutatingwebhookconfiguration is present in the API. If it does not exist,
+// it will be created. If it does exist, the metadata of the required
+// mutatingwebhookconfiguration will be merged with the existing mutatingwebhookconfiguration
+// and an update performed if the mutatingwebhookconfiguration spec and metadata differ from
+// the previously required spec and metadata. For further detail, check the top-level comment.
+func ApplyMutatingWebhookConfiguration(client admissionregistrationclientv1.MutatingWebhookConfigurationInterface, recorder events.Recorder,
+	requiredOriginal *admissionregistrationv1.MutatingWebhookConfiguration, expectedGeneration int64) (*admissionregistrationv1.MutatingWebhookConfiguration, bool, error) {
+
+	required := requiredOriginal.DeepCopy()
+	err := SetSpecHashAnnotation(&required.ObjectMeta, required.Webhooks)
+	if err != nil {
+		return nil, false, err
+	}
+
+	existing, err := client.Get(context.TODO(), required.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		created, err := client.Create(context.TODO(), required, metav1.CreateOptions{})
+		reportCreateEvent(recorder, created, err)
+		return created, true, err
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	// Providing upgrade compatibility with service-ca-bundle operator
+	// and ignore clientConfig.caBundle changes on "inject-cabundle" annotation
+	if required.GetAnnotations()[genericCABundleInjectorAnnotation] != "" {
+		if err := copyMutatingWebhookCABundle(required, *existing); err != nil {
+			return nil, false, err
+		}
+	}
+
+	toWrite := existing.DeepCopy()
+	modified := pointer.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(modified, &toWrite.ObjectMeta, required.ObjectMeta)
+	// there was no change to metadata, the generation was right
+	if !*modified && toWrite.ObjectMeta.Generation == expectedGeneration {
```

Comment:
do these webhook support generation at all?

Reply:
Yes, they do, and the way it would be used should looks similar to this: https://github.com/openshift/machine-api-operator/blob/c9a1f7584b4bd025e1e99444d0f6281c86a18a1a/pkg/operator/sync.go#L169-L177

Unfortunately, the fake client does not support generation changes, so to test this we have e2e tests for our operator (https://github.com/openshift/cluster-api-actuator-pkg/pull/176/files#diff-56cc6cb7098e6f7d33d5d2812f6a1aa3R102-R127) and more specific ones: https://github.com/openshift/machine-api-operator/blob/c9a1f7584b4bd025e1e99444d0f6281c86a18a1a/pkg/operator/sync_test.go#L313-L316, both of which are using the real client.